x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedTimingStatsReporter.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.datafeed;
+
+import org.elasticsearch.xpack.core.ml.datafeed.DatafeedTimingStats;
+import org.elasticsearch.xpack.ml.job.persistence.JobResultsPersister;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.function.Function;
+
+public class DatafeedTimingStatsReporter {
+
+    private final Clock clock;
+    private final JobResultsPersister jobResultsPersister;
+    private DatafeedTimingStats persistedTimingStats;
+    private volatile DatafeedTimingStats currentTimingStats;
+
+    public DatafeedTimingStatsReporter(DatafeedTimingStats timingStats, Clock clock, JobResultsPersister jobResultsPersister) {
+        Objects.requireNonNull(timingStats);
+        this.persistedTimingStats = new DatafeedTimingStats(timingStats);
+        this.currentTimingStats = new DatafeedTimingStats(timingStats);
+        this.clock = Objects.requireNonNull(clock);
+        this.jobResultsPersister = Objects.requireNonNull(jobResultsPersister);
+    }
+
+    public DatafeedTimingStats getCurrentTimingStats() {
+        return new DatafeedTimingStats(currentTimingStats);
+    }
+
+    /**
+     * Executes the given function and reports how much time did the execution take.
+     */
+    public <T, R> R executeWithReporting(Function<T, R> function, T argument) {
+        Instant before = clock.instant();
+        R result = function.apply(argument);
+        Instant after = clock.instant();
```

Comment:
So, this is an interesting predicament. I can understand the reason for wanting "wall clock time" for the entire function, but strictly speaking, this is NOT the search time. The search time is returned in the search payload and could gathered from there. 

Though I also see the benefit of having the entire method wrapped. 

Unsure if the complexity is necessary though. Statistically, both the real search time, and the time spent in this method should have 100% correlation and give the same indication for run time.

Reply:
Done.
Thanks for this suggestion Ben. The interface of DatafeedTimingStatsReporter is much cleaner now.